docs/DeferredSignals.md

```diff
@@ -0,0 +1,245 @@
++ - Host signal handler determines deferred region is done, Still has signal in queue.
++ - Pull signal information off of queue
+# JIT jumps to guest signal handler
+# Hopefully guest calls rt_sigreturn instead of long jumping out.
+# Host PC is back at deferred signal section.
++ Deferred region section causes SIGSEGV #2
++ - Host signal handler determines deferred region is done, No signals in the queue.
+# - mprotect signal page to RW.
+# - Continue execution.
++ #1 <Exit deferred region continues>
+```
+
+### Recursive regions with signal in code emitter.
+This one mostly matches the previous example except the behaviour of deferred signal regions leaving.
+
+In this case, if the thread-local refcount is still >0 on `<Exit deferred region>` then there are two behaviours.
+- On ARM64, it will receive a SIGSEGV but the signal handler will increment PC by one instruction and continue execution
+   - Expectation is that signals are significantly less common than `<Exit deferred region>` so the cost of SIGSEGV+PC increment is faster.
+- On x86-64, the region exit checks the refcount before doing the fault access.
+   - This adds more instructions so is slower on average.
+
+This has the expectation that recursive deferred regions both aren't very deep (usually only nested a couple times), and that signals are rare.
+This way there aren't many SIGSEGV checks generated and the signal is finally only handled when reaching the top-most deferred region exit routine.
+
+```
+- #1 <Enter Deferred region - 3 instructions>
+! Compiling JIT Code
+-    #2 <Enter Deferred region - 3 instructions>
+!    Memory allocation
+#    <Async signal received logic from above>
++    #2 <Exit deferred region - 5 instructions>
++    #2 Exit deferred region causes SIGSEGV
++    - TLS refcount is still 1 (from #1)
++    - PC is incremented by one instruction, signal still unhandled.
++ #1 <Exit deferred region - 5 instructions>
++ #1 Exit deferred region causes SIGSEGV
++ <Regular deferred region handling from above called>
+```
+
+### Multiple signals in signal-deferring region
+This is slightly different from the previous iterations since multiple signals in the stack result in odd behaviour.
+
+```
+- #1 <Enter Deferred region - 3 instructions>
+! Compiling JIT Code
+# #1 Asynchronous Signal received
+#   - Signal queued logic
+# #2 Asynchronous Signal received
+#   - Signal queued logic
++ #1 <Exit deferred region - 5 instructions>
++ Exit deferred region causes SIGSEGV
++ <Regular deferred region handling from above called>
++ Guest calls rt_sigreturn
++ - rt_sigreturn handler checks for number of queued signals
+# - mprotect signal page to NONE because signals is > 0
+# - JIT is back to #1 <Exit deferred region>
+# - Exit deferred region causes SIGSEGV again.
+# - Regular handler loop occurs
+```
+
+### mcontext_t handling?
+- TODO: Overwrite mcontext_t things? I don't think this matters but there will be some private state that might leak SIGSEGV information?
```

Comment:
What's left to be done for this TODO? If this needs elaborating (as the TODO syntax suggests), it should be elaborated. If it's an implementation TODO, this should be phrased as an actual paragraph.

Reply:
Changed the comment to a future task.

Reply:
Removed so it doesn't matter.